adrs/0003-testing-strategy.md

```diff
@@ -0,0 +1,49 @@
+# Stacks Testing Strategy
+
+**Date:** 2022-11-28
+
+**Related PR:** [#1195](https://github.com/StackExchange/Stacks/pull/1195)
+
+## Issue
+We currently don't test our Stacks components except from manually/ad-hoc executing some visual regression tests of entire documentation pages with BackstopJS. Linting and formatting despite being setup are also often ignored because they are not running methodically in a pipeline.
+
+**A lack of testing gives us very little confidence refactoring and evolving the codebase.**
+
+## Decision
+
+Introduce 2 initial categories of tests for Stacks to increase confidence working on new and existing components:
+- Component/Unit Tests
+- Component Visual Regression Tests
+
+General guidelines:
+- Tests run in a headless browser context ([wtr + playwright](https://modern-web.dev/docs/test-runner/overview/)) to provide higher level of confidence
+- DOM Querying and Events Simulation are done as much as possible with [testing-library](https://testing-library.com/) and [user-event](https://testing-library.com/docs/user-event/intro/) instead of reaching out directly to DOM APIs
+- Every component should be accompanied by related component and visual regression tests (atomic classes are excluded)
+
+### Component/Unit Tests
+These tests should provide full coverage of the component being tested. Happy path, edge and error cases should be all covered.
+
+Guidelines:
+- Tests should not reach for internal implementation/methods but rather tests the component through its public APIs (from the point of view of a consumer)
+- Complex DOM-agnostic business logic can be extracted in separate functions which can be unit tested in isolation
+
+### Component Visual Regression Tests
```

Comment:
What visual regression library do you propose to use here? I'd personally like use to cease use of backstop full-stop. Internally, the product teams use [mabl](https://www.mabl.com/), which does have support for visual regression tests.

Reply:
I would prefer to use a lightweight option to start with which allows us to keep the tests checked in the source code. I also believe it is more important to have a visual regression suite for each individual components instead of taking screenshots of entire documentation pages.
I would recommend to start using [this WTR plugin](https://www.npmjs.com/package/@web/test-runner-visual-regression) as a first step and see how it goes. ([we won't be alone using this approach](https://github.com/adobe/spectrum-web-components/blob/899b6eb36206cea9c61a890ba304136616b4cd71/web-test-runner.utils.js#L13)).
I have already setup a working example in [this PR](https://github.com/StackExchange/Stacks/pull/1194). It is also integrated in our new CI workflow.

Perhaps we could give mabl a try if we want to keep regression tests for the documentation pages (exactly what backstop is doing today). Personally I consider those tests a bit too broad though. I would prefer to invest on other categories such as a11y automated testing.

Reply:
Resolving this conversation since I have been given 2 thumbs up. 🙂